Protocol: make `supported_sync_modes` a required not empty list on `AirbyteStream`

alafanechere opened the pull request:
Closes https://github.com/airbytehq/airbyte/issues/17563

## What
#14547 raised the awareness that the `supported_sync_modes` field in `AirbyteStream` is not a required field and it can be empty. This can lead to unexpected UI behavior as described in the issue. 
We added a  test to SAT (https://github.com/airbytehq/airbyte/pull/14800) to ensure the `supported_sync_mode` is not an empty list, but this is something that can be enforced at the protocol level.

## How
* Change [airbyte_protocol.yaml](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L221) to make the `supported_sync_mode` field of `AirbyteStream` a **required not empty array**.
* Update SAT and CDK accordingly.
* Make changes on impacted source connectors (`source-google-sheet` and `source-azure-table`).

## Suggested review order

1. [airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml](https://github.com/airbytehq/airbyte/pull/15591/files#diff-23f6bc12d2cf0cbb40771b96abb371d79d04b0eeccfe92d276f314325600b029)
2. [airbyte-cdk/python/airbyte_cdk/models/airbyte_protocol.py](https://github.com/airbytehq/airbyte/pull/15591/files#diff-a22224a3797701e1a0a0e8e53eed63c83ebae312c2c174dbbf5a4cbbbca0739c)

## 🚨 User Impact 🚨
I'm aware that protocol updates are critical. I did not file a tech spec as I'm under the impression this modification is a fix and not a protocol evolution. @cgardens mentioned in this [comment](https://github.com/airbytehq/airbyte/blob/master/airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml#L227) that this change can happen *once sources are migrated*. I'd love to have more context about this comment to evaluate the potential impact of this change.

evantahler commented on airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:
```diff
@@ -224,8 +224,7 @@ definitions:
       - json_schema
-      # todo (cgardens) - make required once sources are migrated
-      # - supported_sync_modes
+      - supported_sync_modes
```
I guess we've migrated!

evantahler commented on airbyte-protocol/protocol-models/src/main/resources/airbyte_protocol/airbyte_protocol.yaml:
```diff
@@ -224,8 +224,7 @@ definitions:
       - json_schema
-      # todo (cgardens) - make required once sources are migrated
-      # - supported_sync_modes
+      - supported_sync_modes
```
I guess we've migrated!  Can you confirm that there are no old source that don't provide this property?

alafanechere replied:
I'm not 💯 of the Singer compatibility...

alafanechere replied:
> Can you confirm that there are no old source that don't provide this property?
We can double check that existing GA/beta connectors SAT is passing - a test to check the presence of this property was shipped a week ago #14800

evantahler replied:
I'm going to add @jdpgrailsdev as well to confirm, but I think this is 👍

girarda commented on airbyte-cdk/python/unit_tests/singer/test_singer_source.py:
```diff
@@ -95,7 +95,7 @@ def test_singer_discover_metadata(mock_read_catalog):
     _roles_stream = airbyte_catalog.streams[1]
 
-    assert _user_stream.supported_sync_modes is None
+    # assert _user_stream.supported_sync_modes is None
```
should this assert the right modes are supported?

alafanechere replied:
I'm not sure why this assertion was declared. There is probably an edge case with singer and supported sync modes that I'm unaware of.

girarda reviewed (approved):
:shipit:

alafanechere commented:
This PR will require:
1. Publishing changed made an `source-google-sheet` and `source-azure-table`
2. Publishing SAT
3. Publishing CDK

alafanechere commented:
/test connector=connectors/source-google-sheet

alafanechere commented:
/test connector=connectors/source-azure-table

alafanechere commented:
/test connector=connectors/source-azure-table

alafanechere commented:
/test connector=connectors/source-azure-table

alafanechere commented:
/test connector=connectors/source-e2e-test

jdpgrailsdev reviewed (approved):
:shipit:

alafanechere commented:
/publish-cdk dry-run=true

alafanechere commented:
/publish connector=connectors/source-e2e-test

alafanechere commented:
/publish connector=connectors/source-e2e-test

alafanechere commented:
Platform build did not pass because of acceptance tests failures:
* They rely on an old version of the source-e2e-test connector, which is not respecting my protocol change
* I published the new version of source-e2e-test  and updated the hardcoded version [here](https://github.com/airbytehq/airbyte/blob/de0da4bc8388054fbaba0c264a7dd32f13325c0a/airbyte-test-utils/src/main/java/io/airbyte/test/utils/AirbyteAcceptanceTestHarness.java#L129)
* I updated the config passed to the connector [here](https://github.com/airbytehq/airbyte/blob/de0da4bc8388054fbaba0c264a7dd32f13325c0a/airbyte-tests/src/acceptanceTests/java/io/airbyte/test/acceptance/AdvancedAcceptanceTests.java#L259)

Hopefully they'll pass now 🙏

alafanechere commented:
/publish connector=connectors/source-azure-table,connectors/source-google-sheet

alafanechere commented:
/publish connector=connectors/source-google-sheets

alafanechere commented:
/publish connector=connectors/source-azure-table

alafanechere commented:
/publish connector=connectors/source-azure-table

alafanechere commented:
/publish connector=connectors/source-azure-table

alafanechere commented:
/publish connector=connectors/source-azure-table

alafanechere commented:
/publish connector=connectors/source-azure-table

alafanechere commented:
/publish-cdk dry-run=true

alafanechere commented:
/publish-cdk dry-run=false

alafanechere commented:
/publish-cdk dry-run=false

alafanechere commented:
/publish connector=bases/source-acceptance-test auto-bump-version=false

alafanechere commented:
/test connector=connectors/source-instagram

alafanechere commented:
/test connector=connectors/source-hubspot

alafanechere commented:
/test connector=connectors/source-instagram

alafanechere commented:
/test connector=connectors/source-amazon-ads

alafanechere commented:
/test connector=connectors/source-bing-ads

alafanechere commented:
/test connector=connectors/source-amplitude

alafanechere commented:
/test connector=connectors/source-facebook-marketing

alafanechere commented:
/test connector=connectors/source-instagram

alafanechere commented:
/test connector=connectors/destination-amazon-sqs

alafanechere commented:
/test connector=connectors/destination-firebolt

alafanechere commented:
/test connector=connectors/destination-firestore

alafanechere commented:
/test connector=connectors/destination-rabbitmq

alafanechere commented:
/test connector=connectors/destination-sftp-json

alafanechere commented:
/test connector=connectors/destination-sqlite

alafanechere commented:
/test connector=connectors/destination-kvdb

alafanechere commented:
/test connector=connectors/source-postgres

alafanechere commented:
/test connector=connectors/source-instagram

alafanechere commented:
/test connector=connectors/destination-bigquery-denormalized

alafanechere commented:
/test connector=connectors/destination-cassandra

alafanechere commented:
/test connector=connectors/destination-dynamodb

alafanechere commented:
/test connector=connectors/destination-gcs

alafanechere commented:
/test connector=connectors/destination-jdbc

alafanechere commented:
/test connector=connectors/destination-snowflake

alafanechere commented:
/test connector=connectors/destination-sqlite

alafanechere commented:
/test connector=connectors/source-relational-db

alafanechere commented:
/test connector=connectors/source-relational-db

grubberr commented:
@alafanechere 
https://github.com/airbytehq/airbyte/actions/runs/3279207635/jobs/5398502934#step:11:1089

alafanechere commented:
/test connector=connectors/source-file
